Add `collapseHz()`

brownag opened the pull request:
Here is another draft idea for a common operation people want to perform on SoilProfileCollections. This is a minimal implementation but I think several other options could greatly enhance the ways it could be used. It is likely it could be used to resolve some issues in soilDB such as https://github.com/ncss-tech/soilDB/issues/120, https://github.com/ncss-tech/soilDB/issues/122, https://github.com/ncss-tech/soilDB/issues/135

`collapseHz()` function makes it easy to aggregate adjacent horizons (collapsing them into a single horizon) using groups based on pattern matching. Numeric properties are aggregated via weighted average, and other properties are aggregated via dominant condition (i.e. the aggregate value is the value with greatest cumulative thickness in each group)

TODO/additional ideas 
 - [x] Custom matching function (not necessarily `grepl()` based, like `thicknessOf()` / #306)
 - [x] Custom numeric/categorical aggregation functions
   - [x] Ability to specify multiple summary statistics to be calculated per column (e.g. range, or standard deviation, number of subhorizons, number of NA values, or similar)
   - [x] Concatenation and/or thickness tabulation of categories rather than dominant as the only option
 - [x] Ability to exclude certain numeric properties from numeric aggregation (e.g. color value/chroma should not be averaged)
 - [x] Integrate with generalized horizon labels (via `GHL()`)--in lieu of specified `pattern` and possibly change default for `hzdesgn` when GHL() is set?


## Example behavior

``` r
# testing collapseHz
library(aqp)
#> This is aqp 2.0.3
data(loafercreek, package = "soilDB")

x <- loafercreek[c(5,6,7,10,11),]

# collapse on hzdesgnname
a <- collapseHz(x, c(`A` = "^A", `BA` = "BA|Bw", `Bt` = "[ABC]+t", `Cr` = "Cr|R"))
profile_id(a) <- paste0(profile_id(a), "_collapse")
c(a, x) |> 
  plot(color = "clay")
```

![](https://i.imgur.com/Cp0wmZL.png)<!-- -->

``` r

# collapse on texture class groups / custom hzdesgnname
b <- collapseHz(x, 
                pattern = c(`l` = "^(l|sl)$", `c` = "^(cl|c|sicl|scl)$", `si` = '^si|sil$'),
                hzdesgn = "texcl")
profile_id(b) <- paste0(profile_id(b), "_collapse")
c(b, x) |> 
  plot(color = "texcl", name = "texcl")
```

![](https://i.imgur.com/ORR6goa.png)<!-- -->

dylanbeaudette commented:
This is a great idea, and perfect extension to the GHL foundation of ideas and methods. I'm surprised that we never included it before.

From the texture class example above, it isn't clear to me how a dominant condition can be applied, when there is a (I think) full mapping of new labels to REGEX-matched patterns. Does the dominant condition come into play when an unmatched horizon label is encountered?

Good call on excluding color, perhaps a reasonable place to use simulated mixtures. Or, it could be that someone is only interested in Munsell value, and in that case a wt. mean is probably fine.

brownag commented:
> This is a great idea, and perfect extension to the GHL foundation of ideas and methods. I'm surprised that we never included it before.

Yeah, agreed, its something I feel like we have talked a lot about but just never have implemented a generic function for.

> From the texture class example above, it isn't clear to me how a dominant condition can be applied, when there is a (I think) full mapping of new labels to REGEX-matched patterns. Does the dominant condition come into play when an unmatched horizon label is encountered?

Dominant condition is based on the thickest horizon's value within each "adjacency" group. There is no aggregation applied to an unmatched horizon label--it just passes through to the result. As currently implemented this is modifying the `hzdesgn` column in place, which could be the generalized horizon label column or in these examples is the `hzdesgnname()`.

I need to ponder a bit on this as far as what the right ergonomics are. Probably better to have the function calculate a new horizon designation column and set it as the generalized horizon label, or horizon designation label, or both? Rather than modifying existing data--which might be important for downstream context.

> Good call on excluding color, perhaps a reasonable place to use simulated mixtures. Or, it could be that someone is only interested in Munsell value, and in that case a wt. mean is probably fine.

Yeah, I think I want to keep the function relatively "dumb" but specific columns could be targeted with specific methods by either calling `reduceSPC()` on the input before hand, and then a custom aggregation function (TODO), or alternately an argument that allows exclusion of a vector of some column names from the transformation applied to the horizon data frame, or forces categorical aggregation

brownag commented:
I must note that this concept is implemented in part as `dissolve_hz()` by @smroecker (https://github.com/ncss-tech/aqp/blob/fcad3e8e434b71e6ec61d59e032944a8dc09c24c/R/segment.R#L275)--but that function does not operate on SoilProfileCollection objects. 

I forgot about that function and reinvented part of the wheel here, I suppose. It is nice that `dissolve_hz()` can operate on several grouping variables, which is something that could be considered here as well, but it would behave differently (i.e. it would use the interaction of the specified grouping variables to create adjacent horizon groups, rather than return separate aggregations for each factor)

`collapseHz()` function intent seeks to be more generic (when TODOs are complete), and also does aggregation of properties across the of horizon data.frame in one call, rather than just calculating depths.

dylanbeaudette commented:
A note for later: one of the simplest cases is flattening on an intended horizon index, such as chkey or phiid, with a short-circut for 99% of the cases where no flattening is required.

dylanbeaudette commented on R/collapseHz.R:
```diff
@@ -0,0 +1,279 @@
+      h <- h[order(h[[idn]], h[[hzd[1]]]),]
+      
+      # replace horizons in parent SPC
+      replaceHorizons(x) <- h
```
Should we retain original horizon IDs and set new ones? This level of accounting could be useful, but adds clutter. I'd vote for new horizon IDs and dump the old ones for simplicity.

brownag replied:
> Should we retain original horizon IDs and set new ones? This level of accounting could be useful, but adds clutter. 
I don't think it is a good idea to dump the horizon IDs and calculate new ones. This could create unintended conflicts with the auto-calculated sequential integer IDs.

I think it not desirable for the calculated `hzID` nor one set by the user e.g. `chkey` or `chiid`. Essentially I want to retain reference to the source horizon. I don't see an issue with current behavior that keeps a subset of the parent `hzidcolumn()` values. Can you elaborate on what you are thinking here?

dylanbeaudette replied:
Maybe I missed something: when two horizons are *collapsed* what is the resulting horizon ID?

brownag replied:
The horizon ID (character data type) should be from the thickest horizon in the group. It is not handled differently than any other horizon-level variable by default, but you can supply a custom aggregation function (to e.g. concatenate IDs) if desired

dylanbeaudette reviewed (approved):
Great addition. I'd like to add one more test since there is quite a lot of complexity here. GHL assignment → collapsing horizons is going to be really useful.

brownag commented on R/collapseHz.R:
```diff
@@ -0,0 +1,279 @@
+  # iterate over patterns
+  for (p in seq(pattern)) {
+    
+    h <- data.table::data.table(horizons(x))
```
No, the patterns are applied and inserted into horizon data sequentially.

dylanbeaudette replied:
I see, thanks.

brownag replied:
Actually, I was wrong, I believe this and the `replaceHorizons<-()` bit can be moved outside the loop.

brownag replied:
Done in 424ebbe8

brownag commented on R/collapseHz.R:
```diff
@@ -0,0 +1,279 @@
+      r <- rle(l)
+    }
+    
+    if (any(r$lengths > 1)) {
```
Done in 00980105

brownag commented:
> Great addition. I'd like to add one more test since there is quite a lot of complexity here. GHL assignment → collapsing horizons is going to be really useful.

Thanks. What else would you like to test?

dylanbeaudette commented:
> > Great addition. I'd like to add one more test since there is quite a lot of complexity here. GHL assignment → collapsing horizons is going to be really useful.
> 
> Thanks. What else would you like to test?

Some degenerate cases, and an additional hand-verified example. I can add those after it is merged.

brownag commented:
> > > Great addition. I'd like to add one more test since there is quite a lot of complexity here. GHL assignment → collapsing horizons is going to be really useful.
> > 
> > 
> > Thanks. What else would you like to test?
> 
> Some degenerate cases, and an additional hand-verified example. I can add those after it is merged.

I've tested this on some `fetchNASIS()` results with multiple labsampnum per phiid using `by="phiid"`. Go ahead and add the tests to this PR, I'm not in a rush to merge it if there are other items we still need to check.

dylanbeaudette commented:
Testing out PR, and I see that many numeric columns are converted to character:

```r
data("jacobs2000", package = "aqp")
x <- jacobs2000

a_pattern <- c(`A` = "^A",
               `E` = "E", 
               `Bt` = "[B]+t", 
               `Bh` = "[B]+h", 
               `C` = "^C", 
               `foo` = "bar")

m <- collapseHz(jacobs2000, pattern = a_pattern)
profile_id(m) <- paste0(profile_id(m), "_c")

str(horizons(m))
```

```
'data.frame':	28 obs. of  18 variables:
 $ id                   : chr  "92-1_c" "92-1_c" "92-1_c" "92-1_c" ...
 $ name                 : chr  "A" "E" "Bt" "B/C" ...
 $ sand                 : chr  "85" "80" "51" "41" ...
 $ clay                 : chr  "7" "6" "41" "50" ...
 $ matrix_color         : chr  "#675F56FF" "#B1904AFF" "#BF8B37FF" NA ...
 $ time_saturated       : chr  "0" "0" "0" "0" ...
 $ top                  : num  0 18 43 153 156 0 18 46 145 0 ...
 $ bottom               : num  18 43 153 156 213 18 46 145 213 25 ...
 $ matrix_color_munsell : chr  "10YR 4/1" "2.5Y 6/6" "10YR 6/8" NA ...
 $ depletion_code       : chr  NA NA "f2" NA ...
 $ depletion_munsell    : chr  NA NA "10YR 7/3" NA ...
 $ depletion_color      : chr  NA NA "#C4AC8CFF" NA ...
 $ concentration_code   : chr  NA NA "f2" NA ...
 $ concentration_munsell: chr  NA NA "2.5YR 4/8" NA ...
 $ concentration_color  : chr  NA NA "#964C26FF" NA ...
 $ depletion_pct        : chr  "0" "0" "0" "0" ...
 $ concentration_pct    : chr  "0" "0" "2" "0" ...
 $ hzID                 : chr  "1" "2" "4" "6" ...
```

This only happens sometimes:
```r
# no aggregation required, data are left as-is
m <- collapseHz(jacobs2000, by = 'name')

# numeric data converted to character 
m <- collapseHz(jacobs2000, by = 'matrix_color_munsell')
```

dylanbeaudette commented:
Additional tests I would like to add:
 * checking hand-calculated wt. means with and without NA, 
 * selection of most common category with and without NA
 * degenerate horizon logic -- maybe we error and ask user to fix first
 * missing values in `by`

brownag commented:
> Testing out PR, and I see that many numeric columns are converted to character:

Good catch!! Fixed in 489d2a81  

There were two things: a typo/not replaced variable name that changed last night, which caused the numeric aggregation to not work at all, and also there was some implicit conversion to matrix going on (where first column of matrix was character so all others were) that I also introduced last night. 

The reason it only happens sometimes is the "aggregation" code was not triggered if all of the labels are unique within profiles, essentially the "short circuit" you refer to here: https://github.com/ncss-tech/aqp/pull/307#issuecomment-2116238270

brownag commented:
> Additional tests I would like to add:
> 
> * checking hand-calculated wt. means with and without NA,
> * selection of most common category with and without NA
> * degenerate horizon logic -- maybe we error and ask user to fix first
> * missing values in `by`

Thanks, I added tests and checks for averages and dominant condition with and without NA, and missing values in `by` in https://github.com/ncss-tech/aqp/pull/307/commits/5bcac0734a6b941065f0c34e630b8e29c0c96bb4, https://github.com/ncss-tech/aqp/pull/307/commits/268497df13e5151ffc075beda411e7a6306d091c, https://github.com/ncss-tech/aqp/pull/307/commits/1a848832eb4c6b8b6b70d06eccc786b59a5f1749

With the dominant condition aggregation, rather than thickest, `NA` values were always implicitly removed, though depths/thicknesses were handled. That should be fixed now.

I think I would rather not implement tests of horizon logic within this function, or at least have the option to run it without checking. I think the aggregation will generally work "correctly" even with missing depths, overlaps, gaps, though the results may not be sensible. I added some tests for the a couple common cases (filled profiles, 0 profile SPC, `by` column not in horizon data.frame) in 2902a8a8

One might use something like this function to fix certain types of logic errors by crafting either a pattern or column indicating things that should be combined.

brownag commented:
80dc40f9 adds a basic screening function that just checks for any NA top or bottom depths and for bottom depths shallower than top. A warning directs the user to use `checkHzDepthLogic()` (same as `depths<-()` method)

dylanbeaudette commented:
Looks good to me, merge away!
